### ISSUE TITLE
git history

### ISSUE THREAD

**Issue by lukasmerten** [CRPropa/CRPropa3#224]
Hi all, 
it seems that @TobiasWinchen's fix of amount of contributed code introduces some issues in the git history.

When I try to merge the current CRPropa/master branch into my local master branch I to get a bunch of merge conflicts. Since this fix my master and the CRPropa3/master are diverged by ~1300 commits.

This is probably also related to the huge number of changed files in Mario's pull request  #220.

At least @reichherzerp and @antoniusf can confirm this problem.

I think we should try to resolve this before further more people merging the current CRPropa3/master which might introduce new merge conflicts for future pull-requests.

Best, Lukas

**Comment by antoniusf** [comment]
Hi all,

I've just taken a closer look at this, and it appears to me that the fix has rewritten a large part of the entire commit history (compare the 1300 commits divergence in most forks vs about 1500 commits total). The last non-rewritten commit appears to be at 6f4b17e3f123d636036b43fa056841dd178495aa.

To remove Mike's unnecessary contribution, I do feel like it should be sufficient to to rewrite only @reichherzerp's branch pre-merge, along with the merge commit onto master (e5de0d444542f72f2be3edbb8d2ab3d1101cc0d9), and the three master commits that followed it. While this might still cause problems, it would only affect those who have already pulled in that merge commit.

Of course, this would require to first undo the current fix, which I personally do think is advisable, but it is probably good to discuss this first. I should also say that I find `git filter-branch` to be quite scary; it was probably a lot easier for me to propose the above solution than it would be for someone to implement it.

 -- Antonius

**Comment by TobiasWinchen** [comment]
I guess that rebasing the branches onto master should do the trick

**Comment by lukasmerten** [comment]
I just created a new branch from my current master to test the rebase.
`git checkout -b TestMaster`
Then I tried to rebase on current CRPropa3/master
`git rebase upstream/master`, 
where upstream is pointing to https://github.com/CRPropa/CRPropa3

It did not work and creates a lot of merge conflicts, too.

**Comment by TobiasWinchen** [comment]
@lukasmerten - I just tried to reproduce the problem and cloned your master from github and added CRPropa3/master as upstream:
 `$git remote add upstream https://github.com/CRPropa/CRPropa3 && git fetch upstream`
I then tried to merge 
`$git merge upstream/master`
yielding only few conflicts:

```$git status                                                 
On branch master
Your branch is up to date with 'origin/master'.

You have unmerged paths.
  (fix conflicts and run "git commit")
  (use "git merge --abort" to abort the merge)

Changes to be committed:

        new file:   include/crpropa/module/PropagationBP.h
        new file:   src/module/PropagationBP.cpp

Unmerged paths:
  (use "git add <file>..." to mark resolution)

        both added:      .mailmap
        both modified:   CMakeLists.txt
        both added:      CONTRIBUTING.md
        both added:      doc/Doxyfile.in
        both added:      include/CRPropa.h
        both added:      python/2_headers.i
        both modified:   test/testPropagation.cpp
```
Did I reproduce the state of your repository?

**Comment by lukasmerten** [comment]
Yes that reproduces the error.

Since I am quite busy with writing down my thesis I do not have the time to take a closer look at this problem right now. So I will merge for now and simply always choose the CRPropa3/master file. 

However, I will appreciate it if we avoid rewriting the history on CRPropa3/master in the furture.

**Comment by antoniusf** [comment]
Hey everyone,

This thread appears to have gone a bit quiet in the past few days, but I must admit that I am not quite happy with the state the repository is currently in.

I believe that the right thing to do is to undo the entirety of the fix, and to restore the official master branch to its original un-modified history. In the rest of this post, I will elaborate on my reasoning on this, and I would like to encourage everyone to think about the possible options at this point.

@lukasmerten advised me to tag @adundovi in this discussion, who I hope may have some useful input on this.

## What happened
_Feel free to skip this section if you are already familiar with the details._

Two weeks ago, PR #222 (containing the new Boris Push implementation) was merged into the CRPropa master. Due to an erroneous commit (646422dbbfc8909542419c986ade5182623ba7e0) in the commit history of that branch, @GilbertWilbert subsequently appeared in the contributors overview as having added more than 200,000 lines of code. The commit accidentally included extra data and header files, which are normally separate from the CRPropa source code. These extraneous files were removed in a later commit and thus never had any impact on the official master. Their only effect was an incorrectly large number on the contributors page.

To fix this problem, @TobiasWinchen [used the `git filter-branch` command](https://github.com/CRPropa/CRPropa3/pull/222#issuecomment-464059349) to modify the problematic commit. Due to the way that git works, modifying a single commit necessitates the modification of all commits depending on it – in this case, it would have required modification of the Boris Push branch pre-merge, the merge commit, and the three commits following the merge on the master branch.

It appears that, in the process of doing this, more than 1300 commits dating all the way back to March 2012 were re-written and are now incompatible with almost all existing forks of the repository.

I would like to argue that this change is quite undesirable and that it should be undone.

## Problems with the rewritten history
To save you some time, I'll start with a brief overview of the key points. For those curious about the details, I'm following that up with a more in-depth elaboration on each point.

1. Rewriting the master branch of a public repository may be considered bad practice. Adapting to this new history requires entirely avoidable and useless work on the part of every contributor.
2. The rewritten history is missing at least one commit that is present in the original.
3. Merging the rewritten master into each fork will likely create a somewhat messy commit history.

### 1. Rewriting the master branch may be considered bad practice.
I was chatting with some software developers last weekend and told them about this situation. All of them seemed to be in quite strong agreement that re-writing the master branch of a public repository is really not something that should happen. On the technical side, it messes with git's ability to keep track of branch-off points and the like, and it creates extra work for every active fork. On a more abstract level, it seems to me that the history of the master branch is something that is expected to be reliable, and that changing it should only happen in extraordinary circumstances (such as the removal of sensitive information).

### 2. There is at least one missing commit.
Unfortunately, it isn't just that all of these commits now have a different name. The rewritten version of the commit history appears to be missing at least one commit. (The next paragraph is a bit technical; feel free to skip it if you're not interested.)

The commit [`d4ce2f5`][d4ce2f5], for example, is not present in the rewritten master history. Compare 4082b5808a2d65c622e932f71eae373c61ad8487 from the rewritten history and [`22fa8e3`](https://github.com/lukasmerten/CRPropa3/commit/22fa8e31e91b67b7c8d8eae8c93f03e230b9fd9c) from the original history; they're clearly the same commit. Now, at the top of each commit page, you'll find a small link to the commit's parent. In the rewritten history, that parent commit is 6f4b17e3f123d636036b43fa056841dd178495aa. In the original history, the parent is [`d4ce2f5`][d4ce2f5]. Only when we go to [`d4ce2f5`][d4ce2f5]'s parent is when we find ourselves back at 6f4b17e3f123d636036b43fa056841dd178495aa. In other words: the commit is missing.

Now, that missing commit may not be of particular importance, because it did not, in fact, introduce any changes. However, in git, each commit guarantees the correctness of all commits that came before it. This one small omission is therefore enough to break git's system for all commits that follow, thus causing the 1300 commit difference.

Note that this makes it hard to tell if anything else has gone wrong. If two commits have the same number, they are not only guaranteed to introduce the same changes, but they are also guaranteed to have the same history. If they have different numbers, these guarantees are broken; any number of changes may have occured between their histories.

### 3. Merging will create a messy commit history.
Even though some commits may be missing, most of the rewritten history consists of the exact same commits as the original: Same author, same commit message, same changes. However, because their histories are different, git sees these as different commits (git identifies each commit by a unique number, which is also depended on the commit history). When a fork containing the original history is now merged into master (or the other way around), git just sees that the commit histories branch at 6f4b17e3f123d636036b43fa056841dd178495aa, as this is the last unmodified commit. In principle, both branches contain the same commits, but git doesn't know this. Therefore, after merging, the joint commit history will contain each commit twice: Once from the original history, and once from the rewritten one. To see what this would look like, take a look at the commit history of the [`merged` branch here](https://github.com/antoniusf/CRPropa3/commits/merged?after=1939875be4a2332661169ef354b36a2e5f92e953+34). (Note that this is not a proper merge, since I didn't take the time to resolve all conflicts correctly. It is just meant to show the resulting commit history, which is independent of the way that conflicts are resolved.)

## It is possible to fix this.
The original history is still preserved in most of the forks. Most forks will miss some of the last commits, but those can be included from the rewritten history. This assumes that those commits were unaffected, but if they are not present in any of the forks, we have no way to tell – and neither does git. It would still resolve pretty much all of the above points, and it is the best solution I can think of.

If this turns out to be the desired action, I'm offering to put in the work necessary for reconstructing the original history, which I don't expect to be too difficult. I'd then publish the original history on my repository, so that everyone would be able to check that it is consistent with their histories. It should also be consistent with the files currently in this repository – no source code should be lost, and all changes should be preserved.

## Conclusion

I hope that I've been able to give you an insight into the current situation and the problems that it may create, and I look forward to hearing on your opinions on this matter. If this post has left you with questions, please feel free to ask me or any of the other people here. I appreciate that you're taking the time to think about this.

[d4ce2f5]: https://github.com/lukasmerten/CRPropa3/commit/d4ce2f516b8d693738f3957f8ead5ad10f044977

**Comment by lukasmerten** [comment]
Thanks @antoniusf for summarizing the problem in depth. It explains pretty much what I was thinking about the issue but had neither the time nor the abilities to bring to the point.

In my opinion we should go for the extra work to undo the changes and repair the the git history -- especially when @antoniusf is offering to do that work.

**Comment by adundovi** [comment]
My local repo was synchronised last time in September ( 2071e32 to be more specific), thus, I'm a bit out of context and cannot be too helpful in restoring the most recent healthy state of the original repository. As far as understood the problem, I vote for restoring the original history as @antoniusf suggested. But please, before you proceed make a few backups, just in case :-)

Also, I see that PR #222 should have been squashed and/or rebased before merging. We will know better next time :-)

**Comment by TobiasWinchen** [comment]
I agree with @antoniusf suggestion and apologize for the troubles my quick solution caused - to prevent such things in the future, it might be a good idea to disallow force pushing to master.

**Comment by antoniusf** [comment]
Hi folks,

thanks so much for all of your responses! I really do appreciate that you've taken the time to do this, even though there are probably many other things you'd rather be working on instead.

I have good news for you all: I've started working on the reconstruction, but then realized that the original master commit (and its entire history) might still be lying around in the repository. Luckily, this appears to be the case! Here it is: 0c3b412b760a5feb93f0617f8c59424f25c89c72.

Getting this commit was a bit of a journey, but GitHub's event logger helped me out eventually. I've also confirmed that this commit can be reconstructed (mostly, except for signatures) from the current history.

Note that this commit represents the actual original history: Even @GilbertWilbert's problematic commit is still in it! You can also check that [its commit history](https://github.com/CRPropa/CRPropa3/commits/0c3b412b760a5feb93f0617f8c59424f25c89c72) matches with the one present in the forks. It also contains a few "Verified" commits, which `filter-branch` is not able to reproduce.

This leaves us with the following choice: Either, we restore the master branch's original state (ie 0c3b412b760a5feb93f0617f8c59424f25c89c72), before any modifications. Or, we restore it to a state that is still mostly untouched, but that does not include the problematic commit. As far as I can tell, none of the GitHub forks have pulled in the commits that would be affected by this. Despite this, rewriting that commit is still somewhat icky, and I can't say with confidence that it won't cause any more problems. (The one place it would clash for sure is with @reichherzerp's history of the borisPush branch. Since development on that is finished, though, it _probably_ isn't that problematic.)

If we decide to go for the first option, the master branch can simply be restored from 0c3b412b760a5feb93f0617f8c59424f25c89c72, though appropriate precautions should of course be taken. If someone other than @TobiasWinchen ends up doing the restore, they'd face the problem of not having the old commit locally, as git does not synchronize unreachable commits. In that case, they'd have to use the web interface to create a new branch there, and then pull down that new branch. Feel free to ask if you want some more info on this, or pull down the [`original-master`](https://github.com/antoniusf/CRPropa3/tree/original-master) from my fork, where I've done exactly that.

Other than this, I second @TobiasWinchen's suggestion of protecting the master branch. It seems like a smart idea to me.

**Comment by antoniusf** [comment]
To speed things up a little, I'm just going to suggest right now that we go with the first option, restoring the master branch from its original state. At this point, it seems like the safest, fastest, and simplest option.

While this _will_ mean that the contributors page retains some inaccurate information, I doubt that we're the only project who's had to handle something like this. To be honest, I doubt that anyone will care that much.

@lukasmerten has offered to do the restore together with me, since two pairs of eyes are better than one. Barring a veto from anyone here, I think we'd do this sometime tomorrow or Friday. We'll backup the repository before we start, so we have a recovery option in case something goes wrong. For the time being, we could also preserve the current state of the rewritten master in a separate branch, so that it'll remain available for a bit.

Once we've done this, force push protection can be enabled for the master branch. If there is the need for further discussion on this topic, I think we should do so in a separate issue. That should help keep things tidy, and allow us to wrap this issue up.

**Comment by TobiasWinchen** [comment]
If you are working on this anyway, I would suggest fixing the `bad merge' by squashing the Boris Push PR into fewer commits in one go.

**Comment by antoniusf** [comment]
By "fixing the bad merge", do you mean fixing the inaccurate number of lines added on the contributor overview? This could be achieved relatively simply by combining the original history with the rewritten problematic commit, which would take about five minutes since I already have the scripts set up for this.

I must admit that I dislike performing a `git squash` on the PR in this state, for the following reasons:

1. Squashing these commits is a relatively complex operation. Do we squash everything into one, thus erasing Mike Wilbert's contribution from the commit history? Or do we squash all of @reichherzerp's commits into one and all of Mike's commits into one?

2. In the latter case, the "lines added" number would not actually be fixed, since the extraneous files Mike added were only removed in @reichherzerp's merge commit. We could fix this by changing that commit _in addition to the squash_, but in that case I think it'd be simpler to not do the squash and just rewrite the commit.

3. Squashing commits before merging does not appear to be a policy that is currently enforced on PRs in the repo. I don't think it would be right to retroactively change it for this PR.

Overall, I think that with the current state of the repository, where we are talking about changing the commit history, and are unable to go through the normal PR process, we should keep all modifications as small and simple as possible. Removing some unnecessary files from a commit is one thing, retroactively squashing a PR is, I believe, quite another. If what you care about is just getting that number correct, I think that we should just go for the first option. I do not think that we should do a squash.

Introducing a squash policy on future PRs is something that I think can be discussed separately.

**Comment by TobiasWinchen** [comment]
The idea was to squash out the bogous addition/removal of files only. I still think we should try to fix also the contributors page. So how about reverting the repository to the stage before the merge of the boris push to undo the merge. We can then reapply the (few) changes to master afterwards and clean up the PR branch before a new merge? This is what should have been done in the first place instead of my attempt of a quick fix.

**Comment by antoniusf** [comment]
Oh, ok. I think I just misunderstood what you meant by "squash" in your previous answer. I agree that changing the history in only the borisPush branch is a lot less problematic.

I must admit that I don't quite understand what you're proposing, but what I do get from it seems sound to me. Regarding your repaired master, I'm not quite sure I understand what's going on? Starting with commit d4a9a61125199f739ef421728177948755a95628 (corresponding to commit e9ba9a6254bd97389fdf6dae67c39424bbfc64e8 in the original history), the commits appear to have mismatched hashes. Other commits, like e96c2801a500b12f897117b03dae1cd3dd12efb9, seem to be missing entirely.

Would you mind taking a look at [this version](https://github.com/antoniusf/CRPropa3/commits/restored-master-with-fix)? Originally, it was one of the byproducts of my reconstruction efforts. Its history matches the [original history](https://github.com/CRPropa/CRPropa3/commits/0c3b412b760a5feb93f0617f8c59424f25c89c72) up to the merge commit of the borisPush branch. The merge commit is a little bit different, since it merges in a cleaned up version of the original borisPush branch, excluding the extraneous files. The few following commits to master are unchanged, except for their parent hashes.

To me, this seems to be the thing you're looking for: Roll-back to before the merge, clean up the borisPush branch, then merge the cleaned up version, and re-apply the few commits that follow.

Note that in this version, almost all of the info for each commit is kept the same, including timestamps and authorship info. This may or may not be what you want, since some of these properties appear to have changed in your version.

**Comment by TobiasWinchen** [comment]
> Would you mind taking a look at this version? Originally, it was one of the byproducts of my reconstruction efforts. Its history matches the original history up to the merge commit of the borisPush branch. The merge commit is a little bit different, since it merges in a cleaned up version of the original borisPush branch, excluding the extraneous files. The few following commits to master are unchanged, except for their parent hashes.

This is what I was aiming for with the `RepairedHistory' branch, but I apparently missed some details around the merge.  I will check your branch more closely soon, but so far I think we should go with this version.

**Comment by antoniusf** [comment]
Oh, awesome! Then I'll be curious to hear if it actually matches your intentions after you've inspected it more closely.

**Comment by antoniusf** [comment]
Hi, any updates on this? Should we proceed with the version I posted, or do you need some more time?

**Comment by TobiasWinchen** [comment]
I just checked @antoniousf [branch](https://github.com/antoniusf/CRPropa3/commits/restored-master-with-fix) and I think that we should force push it to master:
 * It contains all history before the merge of the BorisPush branch
 * The BorisPush is more or less cleaned and very similar to my attempt
 * It contains the correct commits added after the merge of the BP Branch

If everybody agrees with this assesment, I can do this tomorrow evening and also add protection for similar thing in the future.

**Comment by reichherzerp** [comment]
I think the branch looks good. Thanks a lot for the work to fix the problems.

**Comment by antoniusf** [comment]
Thanks for taking the time to verify the branch! I hope I wasn't too pushy when I asked for your status.

I'd be very happy if you did the reset! Can't wait to have this off my plate :)

**Comment by TobiasWinchen** [comment]
Done - I will set up branch protection now ...
